include/engine/response_objects.hpp

```diff
@@ -0,0 +1,257 @@
+#ifndef ENGINE_RESPONSE_OBJECTS_HPP_
+#define ENGINE_RESPONSE_OBJECTS_HPP_
+
+#include "engine/guidance/route_step.hpp"
+#include "engine/guidance/step_maneuver.hpp"
+#include "engine/guidance/route_leg.hpp"
+#include "engine/guidance/route.hpp"
+#include "engine/guidance/leg_geometry.hpp"
+#include "engine/hint.hpp"
+#include "engine/polyline_compressor.hpp"
+#include "extractor/turn_instructions.hpp"
+#include "extractor/travel_mode.hpp"
+#include "util/coordinate.hpp"
+#include "util/json_container.hpp"
+#include "util/typedefs.hpp"
+
+#include <boost/assert.hpp>
+#include <boost/optional.hpp>
+#include <boost/range/irange.hpp>
+
+#include <string>
+#include <utility>
+#include <algorithm>
+#include <vector>
+
+namespace osrm
+{
+namespace engine
+{
+namespace detail
+{
+
+std::string instructionToString(extractor::TurnInstruction instruction)
+{
+    std::string token;
+    switch (instruction)
+    {
+    case extractor::TurnInstruction::Continue:
+        token = "continue";
+        break;
+    case extractor::TurnInstruction::BearRight:
+        token = "bear right";
+        break;
+    case extractor::TurnInstruction::TurnRight:
+        token = "turn right";
+        break;
+    case extractor::TurnInstruction::SharpRight:
+        token = "sharp right";
+        break;
+    case extractor::TurnInstruction::UTurn:
+        token = "uturn";
+        break;
+    case extractor::TurnInstruction::SharpLeft:
+        token = "sharp left";
+        break;
+    case extractor::TurnInstruction::TurnLeft:
+        token = "turn left";
+        break;
+    case extractor::TurnInstruction::BearLeft:
+        token = "bear left";
+        break;
+    case extractor::TurnInstruction::ReachedWaypointLocation:
+        token = "waypoint";
+        break;
+    case extractor::TurnInstruction::EnterRoundAbout:
+        token = "enter roundabout";
+        break;
+    case extractor::TurnInstruction::LeaveRoundAbout:
+        token = "leave roundabout";
+        break;
+    case extractor::TurnInstruction::StayOnRoundAbout:
+        token = "roundabout";
+        break;
+    case extractor::TurnInstruction::Depart:
+        token = "depart";
+        break;
+    case extractor::TurnInstruction::Arrive:
+        token = "arrive";
+        break;
+    default:
+        BOOST_ASSERT("Unsupported instruction reached");
+        break;
+    }
+    return token;
```

Comment:
Hmm it's ugly to do the translation here. What about this (concept, untested):

    #define TURNS \
      X(Arrive, "arrive") \
      X(Depart, "depart")

and then

    #define X(sym, repr) sym

    enum class TurnInst {
        TURNS
    };

    #undef X

and for the representation

    #define X(sym, repr) case sym: return repr;
    switch (instruction) {
        TURNS
    }
    #undef X

This trick is called X-macro. It's a hack, you probably shouldn't use it too often. And I love it.